lib/snippets/deploy-to-gke

```diff
@@ -0,0 +1,147 @@
+#!/usr/bin/env ruby
+
+# Usage: deploy-to-gke <app's namespace> <gcloud deployment> <gcloud project>
+
+# Requires the following to be available on Shipit machine:
+#  - gcloud binary available on the shipit machine's path
+#  - kubectl binary available in the shipit machine's path
+#  - $GCLOUD_CREDENTIALS_DIR/my-gcloud-project-name.json must exist
+
+# Optionally, the following variables can be used to override script defaults:
+#  - K8S_TEMPLATE_FOLDER: location of Kubernetes files to deploy. Default is config/<environment>/.
+
+require 'open3'
+require 'securerandom'
+require 'erb'
+require 'json'
+require 'yaml'
+require 'shellwords'
+require 'tempfile'
+
+class GKEDeployment
+  class FatalDeploymentError < StandardError; end
+
+  def initialize(namespace:, gcloud_deployment:, gcloud_project:, environment:, current_sha:, template_folder: nil)
+    @namespace = namespace
+    @gcloud_deployment = gcloud_deployment
+    @gcloud_project = gcloud_project
+    @current_sha = current_sha
+    enforce_required_parmas
+    @id = current_sha + "-#{SecureRandom.hex(4)}"
+    @template_path = './' + (template_folder || "config/deploy/#{environment}")
+
+    raise FatalDeploymentError, "Project config missing at #{ENV['GCLOUD_CREDENTIALS_DIR']}/#{@gcloud_project}.json" unless File.file?("#{ENV['GCLOUD_CREDENTIALS_DIR']}/#{@gcloud_project}.json")
+  end
+
+  def run
+    authorize_gcloud
+    fetch_clusters.each do |cluster|
+      set_kubectl_cluster(cluster)
+      apply_all_templates
+    end
+  rescue FatalDeploymentError => error
+    print_error(error.message)
+    exit 1
+  end
+
+  def template_variables
+    {
+      'current_sha' => @current_sha,
+      'deployment_id' => @id,
+    }
+  end
+
+  private
+
+  def enforce_required_parmas
+    [@namespace, @gcloud_project, @gcloud_deployment, @current_sha].each do |required_param|
+      raise ArgumentError, "#{required_param} is required" unless required_param && !required_param.empty?
+    end
+  end
+
+  def apply_all_templates
+    unless File.directory?(@template_path)
```

Comment:
Similar to what @casperisfine said, it might be nice to bail earlier if we have nothing to deploy for this reason. Suggestion: we could make a validation method that handles the existing presence checks from `enforce_required_parmas` (there's a typo in "params" there btw), plus the keys file and template path validation. Potentially the presence of at least one yml/erb in the folder as well.